source/common/upstream/upstream_impl.cc

```diff
@@ -505,9 +505,18 @@ Host::CreateConnectionData HostImpl::createConnection(
         socket_factory.createTransportSocket(transport_socket_options, host),
         upstream_local_address.socket_options_, transport_socket_options);
   } else if (address_list.size() > 1) {
-    connection = std::make_unique<Network::HappyEyeballsConnectionImpl>(
-        dispatcher, address_list, source_address_selector, socket_factory, transport_socket_options,
-        host, options);
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.use_config_in_happy_eyeballs") &&
+        cluster.happyEyeballsConfig().has_value()) {
+      ENVOY_LOG(debug, "Upstream using happy eyeballs config.");
+      connection = std::make_unique<Network::HappyEyeballsConnectionImpl>(
+          dispatcher, address_list, source_address_selector, socket_factory,
+          transport_socket_options, host, options, cluster.happyEyeballsConfig().value());
```

Comment:
Now that `happyEyeballsConfig()` returns an optional, can we have just a single constructor which takes this optional and does the initialization conditionally?

Reply:
The reason that I didn't consolidate the constructors is: 1. they advise against passing optional as params [go/totw/163](https://goto.google.com/totw/163). Maybe I can change to ptr? 2. I'm trying to avoid mixing the logics together, if you're not concerned, I'm happy to make the change

Reply:
I think the advice there is: 

"Avoid function parameters of the form const std::optional&."

But it seems to recommend just std::optional<T>. In this case, the T is fairly small so passing by value is fine, I think. (And that's what `cluster.happyEyeballsConfig()` returns anyway.

Reply:
done